# CODSOFT TASK 2
Credit Card Fraud Detection using Machine Learning




In [ ]:
import pandas as pd

from sklearn.model_selection import train_test_split

from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier

from sklearn.metrics import accuracy_score

In [ ]:
data = pd.read_csv("/content/drive/MyDrive/CodSoft DATASETS/fraudTrain.csv")

**Dataset Analysis**

In [ ]:
print(data.head())

   Unnamed: 0 trans_date_trans_time            cc_num  \
0           0   2019-01-01 00:00:18  2703186189652095   
1           1   2019-01-01 00:00:44      630423337322   
2           2   2019-01-01 00:00:51    38859492057661   
3           3   2019-01-01 00:01:16  3534093764340240   
4           4   2019-01-01 00:03:06   375534208663984   

                             merchant       category     amt      first  \
0          fraud_Rippin, Kub and Mann       misc_net    4.97   Jennifer   
1     fraud_Heller, Gutmann and Zieme    grocery_pos  107.23  Stephanie   
2                fraud_Lind-Buckridge  entertainment  220.11     Edward   
3  fraud_Kutch, Hermiston and Farrell  gas_transport   45.00     Jeremy   
4                 fraud_Keeling-Crist       misc_pos   41.96      Tyler   

      last gender                        street  ...      lat      long  \
0    Banks      F                561 Perry Cove  ...  36.0788  -81.1781   
1     Gill      F  43039 Riley Greens Suite 393  ...  48

In [ ]:
print(data.shape)

(1296675, 23)


In [ ]:
print(data.describe())

         Unnamed: 0        cc_num           amt           zip           lat  \
count  1.296675e+06  1.296675e+06  1.296675e+06  1.296675e+06  1.296675e+06   
mean   6.483370e+05  4.171920e+17  7.035104e+01  4.880067e+04  3.853762e+01   
std    3.743180e+05  1.308806e+18  1.603160e+02  2.689322e+04  5.075808e+00   
min    0.000000e+00  6.041621e+10  1.000000e+00  1.257000e+03  2.002710e+01   
25%    3.241685e+05  1.800429e+14  9.650000e+00  2.623700e+04  3.462050e+01   
50%    6.483370e+05  3.521417e+15  4.752000e+01  4.817400e+04  3.935430e+01   
75%    9.725055e+05  4.642255e+15  8.314000e+01  7.204200e+04  4.194040e+01   
max    1.296674e+06  4.992346e+18  2.894890e+04  9.978300e+04  6.669330e+01   

               long      city_pop     unix_time     merch_lat    merch_long  \
count  1.296675e+06  1.296675e+06  1.296675e+06  1.296675e+06  1.296675e+06   
mean  -9.022634e+01  8.882444e+04  1.349244e+09  3.853734e+01 -9.022646e+01   
std    1.375908e+01  3.019564e+05  1.284128e+07  5.

In [ ]:
print(data.isnull().sum())

Unnamed: 0               0
trans_date_trans_time    0
cc_num                   0
merchant                 0
category                 0
amt                      0
first                    0
last                     0
gender                   0
street                   0
city                     0
state                    0
zip                      0
lat                      0
long                     0
city_pop                 0
job                      0
dob                      0
trans_num                0
unix_time                0
merch_lat                0
merch_long               0
is_fraud                 0
dtype: int64


In [ ]:
print(data["is_fraud"].value_counts())

is_fraud
0    1289169
1       7506
Name: count, dtype: int64


In [ ]:
print(data["is_fraud"].value_counts(normalize=True) * 100)

is_fraud
0    99.421135
1     0.578865
Name: proportion, dtype: float64


In [ ]:
data = data.drop("Unnamed: 0", axis=1, errors="ignore")

**Data Preprocessing**

In [ ]:
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()

In [ ]:
data["merchant"] = le.fit_transform(data["merchant"])

data["category"] = le.fit_transform(data["category"])

data["gender"] = le.fit_transform(data["gender"])

data["city"] = le.fit_transform(data["city"])

data["state"] = le.fit_transform(data["state"])

data["job"] = le.fit_transform(data["job"])

In [ ]:
print(data.head())

  trans_date_trans_time            cc_num  merchant  category     amt  \
0   2019-01-01 00:00:18  2703186189652095       514         8    4.97   
1   2019-01-01 00:00:44      630423337322       241         4  107.23   
2   2019-01-01 00:00:51    38859492057661       390         0  220.11   
3   2019-01-01 00:01:16  3534093764340240       360         2   45.00   
4   2019-01-01 00:03:06   375534208663984       297         9   41.96   

       first     last  gender                        street  city  ...  \
0   Jennifer    Banks       0                561 Perry Cove   526  ...   
1  Stephanie     Gill       0  43039 Riley Greens Suite 393   612  ...   
2     Edward  Sanchez       1      594 White Dale Suite 530   468  ...   
3     Jeremy    White       1   9443 Cynthia Court Apt. 038    84  ...   
4      Tyler   Garcia       1              408 Bradley Rest   216  ...   

       lat      long  city_pop  job         dob  \
0  36.0788  -81.1781      3495  370  1988-03-09   
1  48.8878 -11

In [ ]:
data["first"] = le.fit_transform(data["first"])

data["last"] = le.fit_transform(data["last"])

data["street"] = le.fit_transform(data["street"])

data["trans_num"] = le.fit_transform(data["trans_num"])

data["trans_date_trans_time"] = le.fit_transform(data["trans_date_trans_time"])

data["dob"] = le.fit_transform(data["dob"])

In [ ]:
print(data.dtypes)

trans_date_trans_time      int64
cc_num                     int64
merchant                   int64
category                   int64
amt                      float64
first                      int64
last                       int64
gender                     int64
street                     int64
city                       int64
state                      int64
zip                        int64
lat                      float64
long                     float64
city_pop                   int64
job                        int64
dob                        int64
trans_num                  int64
unix_time                  int64
merch_lat                float64
merch_long               float64
is_fraud                   int64
dtype: object


In [ ]:
X = data.drop("is_fraud", axis=1)

y = data["is_fraud"]

In [ ]:
print(X.shape)

print(y.shape)

(1296675, 21)
(1296675,)


**Train-Test Split**

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)

In [ ]:
print(X_train.shape)

print(X_test.shape)

(1037340, 21)
(259335, 21)


In [ ]:
print("Genuine Transactions :", round((data["is_fraud"]== 0).mean()*100,2), "%")
print("Fraud Transactions :", round((data["is_fraud"]==1).mean()*100,2), "%")

Genuine Transactions : 99.42 %
Fraud Transactions : 0.58 %


**Model Training**

**Logistic Regression**

In [ ]:
lr_model = LogisticRegression(max_iter=1000)

lr_model.fit(X_train, y_train)

LogisticRegression(max_iter=1000)

In [ ]:
lr_pred = lr_model.predict(X_test)

In [ ]:
lr_accuracy = accuracy_score(y_test, lr_pred)

print("Logistic Regression Accuracy :", round(lr_accuracy * 100, 2), "%")

Logistic Regression Accuracy : 99.41 %


**Decision Tree**

In [ ]:
dt_model = DecisionTreeClassifier()

dt_model.fit(X_train, y_train)

DecisionTreeClassifier()

In [ ]:
dt_pred = dt_model.predict(X_test)

In [ ]:
dt_accuracy = accuracy_score(y_test, dt_pred)

print("Decision Tree Accuracy :", round(dt_accuracy * 100, 2), "%")

Decision Tree Accuracy : 99.62 %


**Random Forest**

In [ ]:
rf_model = RandomForestClassifier()

rf_model.fit(X_train, y_train)

RandomForestClassifier()

In [ ]:
rf_pred = rf_model.predict(X_test)

In [ ]:
rf_accuracy = accuracy_score(y_test, rf_pred)

print("Random Forest Accuracy :", round(rf_accuracy * 100, 2), "%")

Random Forest Accuracy : 99.77 %


**Accuracy Comparison**

In [ ]:
print("Logistic Regression :", round(lr_accuracy * 100, 2), "%")

print("Decision Tree :", round(dt_accuracy * 100, 2), "%")

print("Random Forest :", round(rf_accuracy * 100, 2), "%")

Logistic Regression : 99.41 %
Decision Tree : 99.62 %
Random Forest : 99.77 %


Conclusion

The fraud detection dataset was analyzed and processed using machine learning techniques. Among all models, Random Forest performed the best with an accuracy of 99.77%.